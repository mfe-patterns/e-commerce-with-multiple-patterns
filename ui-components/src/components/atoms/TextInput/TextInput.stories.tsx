import { storiesOf } from '@storybook/react';
import React from 'react';

import { TextInput } from './TextInput';
import { Wrapper } from './TextInput.styles';

storiesOf('Atoms', module).add('TextInput', () => (
  <>
    <Wrapper>
      <TextInput
        placeholder="Placeholder"
      />
    </Wrapper>
    <Wrapper>
      <TextInput
        success
        placeholder="Placeholder"
      />
    </Wrapper>
    <Wrapper>
      <TextInput
        error
        placeholder="Placeholder"
      />
    </Wrapper>
    <Wrapper>
      <TextInput
        disabled
        placeholder="Placeholder"
      />
    </Wrapper>
  </>
));
