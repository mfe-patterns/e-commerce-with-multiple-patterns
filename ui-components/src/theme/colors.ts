interface Shade {
  50: string;
  100: string;
  200: string;
  300: string;
  400: string;
  500: string;
  600: string;
  700: string;
}

const dark: Shade = {
  50: "#11111193",
  100: "#111111a2",
  200: "#111111B3",
  300: "#111111B9",
  400: "#111111CC",
  500: "#111111d8",
  600: "#111111eb",
  700: "#111111",
};
const neutral: Shade = {
  50: "#fafafa",
  100: "#f2f2f2",
  200: "#ededed",
  300: "#dedede",
  400: "#cccccc",
  500: "#666666",
  600: "#434343",
  700: "#343434",
};

const primary: Shade = {
  50: "#e3f2fd",
  100: "#bbdefb",
  200: "#90caf9",
  300: "#64b5f6",
  400: "#42a5f5",
  500: "#2196f3",
  600: "#1e88e5",
  700: "#1976d2",
};

const white = "#ffffff";

export interface IColorPallete {
  neutral: Shade;
  dark: Shade;
  primary: Shade;
  white: string;
  black: string;
  error: string;
  success: string;
  text: string;
  transparent: string;
  link: string;
}

export const colors: IColorPallete = {
  neutral,
  dark,
  primary,
  white,
  black: "#000000",
  error: "#dc3545",
  success: "#28a745",
  text: dark[700],
  transparent: "transparent",
  link: "#0064cc",
};
