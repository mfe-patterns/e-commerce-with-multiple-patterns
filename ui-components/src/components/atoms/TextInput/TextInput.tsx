import React, { useState } from 'react';
import * as S from './TextInput.styles';

export interface ITextInputProps {
  disabled?: boolean;
  placeholder?: string;
  error?: boolean;
  success?: boolean;
  onFocus?: VoidFunction;
  onBlur?: VoidFunction;
  maxLength?: number;
  height?: number;
}

export const TextInput = React.forwardRef<HTMLElement, ITextInputProps>(
  (props: ITextInputProps, ref) => {
    const {
      disabled = false,
      error = false,
      success = false,
      onFocus,
      onBlur,
      maxLength,
      placeholder,
      ...rest
    } = props;
    const [active, setActive] = useState(false);

    const onInputFocus = () => {
      setActive(true);
      if (onFocus) {
        onFocus();
      }
    };

    const onInputBlur = () => {
      setActive(false);
      if (onBlur) {
        onBlur();
      }
    };

    return (
      <S.StyledTextWrapper>
        <S.StyledTextInput
          maxLength={maxLength}
          error={error}
          success={success}
          active={active}
          placeholder={placeholder}
          disabled={disabled}
          onFocus={onInputFocus}
          onBlur={onInputBlur}
          ref={ref}
          {...rest}
        />              
      </S.StyledTextWrapper>
    );
  },
);
