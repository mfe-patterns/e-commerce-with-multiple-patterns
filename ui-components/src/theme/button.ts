import { colors } from "./colors";

interface IButtonType {
  color: string;
  backgroundColor?: string;
  borderColor?: string;
  fontFamily: string;
  border: string;
  disabled: {
    color: string;
    backgroundColor?: string;
    borderColor?: string;
  };
  horizontalPadding: boolean;
}

export const primary: IButtonType = {
  color: colors.white,
  borderColor: colors.transparent,
  fontFamily: "Roboto-Regular",
  border: "none",
  disabled: {
    color: colors.white,
    borderColor: colors.transparent,
  },
  horizontalPadding: true,
};
export const secondary: IButtonType = {
  backgroundColor: colors.white,
  color: colors.text,
  fontFamily: "Roboto-Regular",
  border: "1px solid",
  disabled: {
    backgroundColor: colors.white,
    color: colors.dark[100],
  },
  horizontalPadding: true,
};

interface IButtonSize {
  height: string;
  borderRadius: string;
  fontSize: string;
  lineHeight: string;
  letterSpacing: string;
  paddingVertical: string;
  paddingHorizontal: string;
}

interface ISize {
  xlarge: IButtonSize;
  large: IButtonSize;
  medium: IButtonSize;
  small: IButtonSize;
  xsmall: IButtonSize;
}

const size: ISize = {
  xlarge: {
    height: "56px",
    borderRadius: "28px",
    fontSize: "16px",
    lineHeight: "24px",
    letterSpacing: "0.7px",
    paddingVertical: "15px",
    paddingHorizontal: "29.5px",
  },
  large: {
    height: "46px",
    borderRadius: "23px",
    fontSize: "15px",
    lineHeight: "20px",
    letterSpacing: "0.7px",
    paddingVertical: "12px",
    paddingHorizontal: "25.5px",
  },
  medium: {
    height: "40px",
    borderRadius: "20px",
    fontSize: "14px",
    lineHeight: "20px",
    letterSpacing: "0.7px",
    paddingVertical: "10px",
    paddingHorizontal: "21.5px",
  },
  small: {
    height: "36px",
    borderRadius: "18px",
    fontSize: "13px",
    lineHeight: "19px",
    letterSpacing: "0.7px",
    paddingVertical: "8px",
    paddingHorizontal: "20px",
  },
  xsmall: {
    height: "29px",
    borderRadius: "14.5px",
    fontSize: "12px",
    lineHeight: "19px",
    letterSpacing: "0.7px",
    paddingVertical: "5px",
    paddingHorizontal: "14px",
  },
};

export interface IButton {
  type: {
    primary: IButtonType;
    secondary: IButtonType;
  },
  size: ISize;
}

export const button: IButton = {
  type: {
    primary: {
      ...primary,
      disabled: {
        ...primary.disabled,
        backgroundColor: colors.primary[200],
      },
      backgroundColor: colors.primary[500],
    },
    secondary: {
      ...secondary,
      borderColor: colors.primary[500],
      disabled: {
        ...secondary.disabled,
        borderColor: colors.primary[200],
      },
    },
  },
  size,
};
