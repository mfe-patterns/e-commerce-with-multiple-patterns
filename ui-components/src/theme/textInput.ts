import { colors } from './colors';
import { FontFamily } from './typography';

interface ITextInputType {
  color: string;
  placeholderColor?: string;
  backgroundColor: string;
  borderColor: string;
}

export interface ITextInput {
  fontFamily: FontFamily;
  default: ITextInputType;
  disabled: ITextInputType;
  error: ITextInputType;
  success: ITextInputType;
}

export const textInput: ITextInput = {
  fontFamily: 'Roboto-Regular',
  default: {
    color: colors.text,
    placeholderColor: colors.dark[100],
    backgroundColor: colors.white,
    borderColor: colors.neutral[300],
  },
  disabled: {
    color: colors.dark[50],
    backgroundColor: colors.neutral[50],
    borderColor: colors.neutral[300],
  },
  error: {
    color: colors.text,
    backgroundColor: colors.white,
    borderColor: colors.error,
  },
  success: {
    color: colors.text,
    backgroundColor: colors.white,
    borderColor: colors.success,
  },
};
