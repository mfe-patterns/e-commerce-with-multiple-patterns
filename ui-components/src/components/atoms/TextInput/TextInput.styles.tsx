import styled, { css } from 'styled-components';
import { theme } from '../../../theme/theme';

type StyledTextInputProps = {
  active: boolean;
  error: boolean;
  success: boolean;
  disabled: boolean;
  enableClearInput?: boolean;
  theme: typeof theme;
  height?: number;
  ref?: any;
};

export const Wrapper = styled.div`
  margin: 20px;
`;

export const StyledTextInput = styled.input<StyledTextInputProps>`
  height: ${({ theme, height }) => `${height || theme.gridSize * 6}px`};
  padding: ${({ theme }) =>
    `${theme.gridSize * 1.5}px ${theme.gridSize * 2}px`};
  color: ${({ theme }) => theme.textInput.default.color};
  background-color: ${({ theme }) =>
    theme.textInput.default.backgroundColor};
  border: 1px solid;
  border-color: ${({ theme }) =>
    theme.textInput.default.borderColor};
  box-sizing: border-box;
  width: 100%;
  
    ${({ active, success, error, theme }) =>
    active &&
    !success &&
    !error &&
    css`
        border-color: ${theme.colors.primary[500]};
      `}
    ${({ success, theme }) =>
    success &&
    css`
      color: ${theme.textInput.success.color};
      background-color: ${theme.textInput.success.backgroundColor};
      border-color: ${theme.textInput.success.borderColor};
      padding-right: ${theme.gridSize * 6}px;
    `}
    ${({ error, theme }) =>
    error &&
    css`
        color: ${theme.textInput.error.color};
        background-color: ${theme.textInput.error.backgroundColor};
        border-color: ${theme.textInput.error.borderColor};
        padding-right: ${theme.gridSize * 6}px;
      `}
    ${({ disabled, theme }) =>
    disabled &&
    css`
        color: ${theme.textInput.disabled.color};
        background-color: ${theme.textInput.disabled.backgroundColor};
        border-color: ${theme.textInput.disabled.borderColor};
      `};
`;

export const StyledTextWrapper = styled.div<any>`
  min-height: ${({ theme }) => `${theme.gridSize * 6}px`};
  justify-content: center;
  flex-grow: 1;
  flex-shrink: 1;
`;
