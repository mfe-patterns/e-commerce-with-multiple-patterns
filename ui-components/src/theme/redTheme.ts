import { ITypography, typography } from "./typography";
import { IColorPallete, colors } from "./colors";
import { ITextInput, textInput } from "./textInput";
import { button, IButton, primary, secondary } from "./button";

const redPalette: IColorPallete = {
  ...colors,
  primary: {
    50: "#ffebee",
    100: "#ffcdd2",
    200: "#ef9a9a",
    300: "#e57373",
    400: "#ef5350",
    500: "#f44336",
    600: "#e53935",
    700: "#d32f2f",
  },
};

const redButton: IButton = {
  ...button,
  type: {
    primary: {
      ...primary,
      disabled: {
        ...primary.disabled,
        backgroundColor: redPalette.primary[200],
      },
      backgroundColor: redPalette.primary[500],
    },
    secondary: {
      ...secondary,
      borderColor: redPalette.primary[500],
      disabled: {
        ...secondary.disabled,
        borderColor: redPalette.primary[200],
      },
    },
  },
};

interface ITheme {
  typography: ITypography;
  colors: IColorPallete;
  textInput: ITextInput;
  button: IButton;
  gridSize: number;
}

export const theme: ITheme = {
  typography,
  colors: redPalette,
  textInput,
  button: redButton,
  gridSize: 8,
};
