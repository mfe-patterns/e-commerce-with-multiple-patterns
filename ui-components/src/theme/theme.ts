import { ITypography, typography } from "./typography";
import { IColorPallete, colors } from "./colors";
import { ITextInput, textInput } from "./textInput";
import { button, IButton } from "./button";

interface ITheme {
  typography: ITypography;
  colors: IColorPallete;
  textInput: ITextInput;
  button: IButton;
  gridSize: number;
}

export const theme: ITheme = {
  typography,
  colors,
  textInput,
  button,
  gridSize: 8
};
