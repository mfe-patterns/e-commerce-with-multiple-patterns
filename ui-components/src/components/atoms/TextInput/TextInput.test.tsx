import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import 'jest-styled-components';

import { TextInput } from './TextInput';
import { theme } from '../../../theme/theme';

describe('TextInput component', () => {
  it('should render with default props', () => {
    const { container, getByTestId } = render(
      <ThemeProvider theme={theme}>
        <TextInput data-testid="text-input" />
      </ThemeProvider>,
    );

    expect(container).toMatchSnapshot();
    expect(getByTestId('text-input')).toHaveStyleRule('color', '#111111');
    expect(getByTestId('text-input')).toHaveStyleRule(
      'background-color',
      '#ffffff',
    );
    expect(getByTestId('text-input')).toHaveStyleRule(
      'border-color',
      '#dedede',
    );
  });

  it('should render with explicit height', () => {
    const { getByTestId } = render(
      <ThemeProvider theme={theme}>
        <TextInput height={100} data-testid="text-input" />
      </ThemeProvider>,
    );
    expect(getByTestId('text-input')).toHaveStyleRule('height', '100px');
  });

  it('should render with error', () => {
    const { getByTestId } = render(
      <ThemeProvider theme={theme}>
        <TextInput data-testid="text-input" error />
      </ThemeProvider>,
    );

    expect(getByTestId('text-input')).toHaveStyleRule('color', '#111111');
    expect(getByTestId('text-input')).toHaveStyleRule(
      'background-color',
      '#ffffff',
    );
    expect(getByTestId('text-input')).toHaveStyleRule(
      'border-color',
      '#dc3545',
    );
  });

  it('renders with success', () => {
    const { getByTestId } = render(
      <ThemeProvider theme={theme}>
        <TextInput data-testid="text-input" success />
      </ThemeProvider>,
    );

    expect(getByTestId('text-input')).toHaveStyleRule('color', '#111111');
    expect(getByTestId('text-input')).toHaveStyleRule(
      'background-color',
      '#ffffff',
    );
    expect(getByTestId('text-input')).toHaveStyleRule(
      'border-color',
      '#28a745',
    );
  });

  it('should trigger onBlur', () => {
    const onBlur = jest.fn();
    const { getByTestId } = render(
      <ThemeProvider theme={theme}>
        <TextInput onBlur={onBlur} data-testid="text-input" />
      </ThemeProvider>,
    );

    fireEvent.blur(getByTestId('text-input'));

    expect(onBlur).toHaveBeenCalled();
  });

  it('should trigger onFocus', () => {
    const onFocus = jest.fn();
    const { getByTestId } = render(
      <ThemeProvider theme={theme}>
        <TextInput onFocus={onFocus} data-testid="text-input" />
      </ThemeProvider>,
    );

    fireEvent.focus(getByTestId('text-input'));

    expect(onFocus).toHaveBeenCalled();
  });
});
