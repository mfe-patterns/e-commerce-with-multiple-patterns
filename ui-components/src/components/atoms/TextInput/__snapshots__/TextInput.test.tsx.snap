// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput component should render with default props 1`] = `
.c1 {
  height: 48px;
  padding: 12px 16px;
  color: #111111;
  background-color: #ffffff;
  border: 1px solid;
  border-color: #dedede;
  box-sizing: border-box;
  width: 100%;
}

.c0 {
  min-height: 48px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
}

<div>
  <div
    class="c0"
  >
    <input
      class="c1"
      data-testid="text-input"
    />
  </div>
</div>
`;
